src/cudadev/CUDACore/getCachingDeviceAllocator.h

```diff
@@ -64,10 +64,24 @@ namespace cms::cuda::allocator {
 
     static cms::cuda::ScopedSetDevice setDevice(DeviceType device) { return cms::cuda::ScopedSetDevice(device); }
 
+    static DeviceType memoryDevice(DeviceType deviceEvent) {
+      // For device allocator the device where the memory is allocated
+      // on is the same as the device where the event is recorded on.
+      return deviceEvent;
+    }
+
     static bool canReuseInDevice(DeviceType a, DeviceType b) { return a == b; }
 
     static bool canReuseInQueue(QueueType a, QueueType b) { return a == b; }
 
+    template <typename C>
+    static bool deviceCompare(DeviceType a_dev, DeviceType b_dev, C&& compare) {
+      if (a_dev == b_dev) {
+        return compare();
+      }
+      return a_dev < b_dev;
```

Comment:
If I understood the "Unified Virtual Addressing" correctly, using the device id explicitly to disambiguate device pointers from different devices should not be necessary on devices that support it (and also the device id can be obtained from the address). The use of `cudaMemcpyDefault` seems to be part of UVA as well, so in practice we're already requiring it. If we can rely on it here, the allocator API could be further simplified (assuming the overhead of `cudaPointerGetAttributes()` is negligible).

Reply:
> The use of `cudaMemcpyDefault` seems to be part of UVA as well, so in practice we're already requiring it.

According to [3.2.8. Unified Virtual Address Space](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#unified-virtual-address-space) UVA requires 64 bit compilation and SM >= 2.0, so I'd say we run very little risks of ending up on any hardware that does not support it.

Reply:
In fact, I think unified addressing is built-in also in HIP and oneAPI.
Can we just consider it a requirement for any technology used with CMSSW ?